File: py/patterns/components/games.py
from .builder import MazeBuilder
from .doors import Door, DoorNeedingSpell
from .factories import MazeFactory
from .mazes import Maze
from .common import Direction
from .rooms import Room, RoomWithABomb, EnchantedRoom
from .spells import Spell
from .walls import Wall, BombedWall


def build(builder: MazeBuilder) -> Maze:
    builder.build_maze()
    builder.build_room(1)
    builder.build_room(2)
    builder.build_door(1, 2)

    return builder.get_maze()


def create(factory: MazeFactory) -> Maze:
    a_maze = Maze()

    r1 = factory.make_room(1)
    r2 = factory.make_room(2)

    d = factory.make_door(r1, r2)

    a_maze.add_room(r1)
    a_maze.add_room(r2)

    r1.set_side(Direction.North, factory.make_wall())
    r1.set_side(Direction.East, d)
    r1.set_side(Direction.South, factory.make_wall())
    r1.set_side(Direction.West, factory.make_wall())

    r2.set_side(Direction.North, factory.make_wall())
    r2.set_side(Direction.East, factory.make_wall())
    r2.set_side(Direction.South, factory.make_wall())
    r2.set_side(Direction.West, d)

    return a_maze


class MazeGame:
    def create(self) -> Maze:
        m = self.make_maze()

        r1 = self.make_room(1)
        r2 = self.make_room(2)

        d = self.make_door(r1, r2)

        m.add_room(r1)
        m.add_room(r2)

        r1.set_side(Direction.North, self.make_wall())
        r1.set_side(Direction.East, d)
        r1.set_side(Direction.South, self.make_wall())
        r1.set_side(Direction.West, self.make_wall())

        r2.set_side(Direction.North, self.make_wall())
        r2.set_side(Direction.East, self.make_wall())
        r2.set_side(Direction.South, self.make_wall())
        r2.set_side(Direction.West, d)

        return m

    def make_maze(self) -> Maze:
        return Maze()

    def make_room(self, n: int) -> Room:
        return Room(n)

    def make_wall(self) -> Wall:
        return Wall()

    def make_door(self, r1: Room, r2: Room) -> Door:
        return Door(r1, r2)


class BombedMazeGame(MazeGame):
    def make_wall(self) -> Wall:
        return BombedWall()

    def make_room(self, n: int) -> Room:
        return RoomWithABomb(n)


class EnchantedMazeGame(MazeGame):
    def make_room(self, n: int) -> Room:
        return EnchantedRoom(n, Spell("game spell"))

    def make_door(self, r1: Room, r2: Room) -> Door:
        return DoorNeedingSpell(r1, r2, Spell('spell game door'))

File: goimp/components/games.go
package components

func CreateGame(factory Factory) *Maze {
	var aMaze = createMaze()

	var r1 = factory.MakeRoom(1)
	var r2 = factory.MakeRoom(2)

	var d = factory.MakeDoor(r1, r2)

	aMaze.AddRoom(r1)
	aMaze.AddRoom(r2)

	r1.SetSide(DirectionNorth, factory.MakeWall())
	r1.SetSide(DirectionEast, d)
	r1.SetSide(DirectionSouth, factory.MakeWall())
	r1.SetSide(DirectionWest, factory.MakeWall())

	r2.SetSide(DirectionNorth, factory.MakeWall())
	r2.SetSide(DirectionEast, factory.MakeWall())
	r2.SetSide(DirectionSouth, factory.MakeWall())
	r2.SetSide(DirectionWest, d)

	return aMaze
}

func BuildGame(builder MazeBuilder) *Maze {
	builder.BuildMaze()
	builder.BuildRoom(1)
	builder.BuildRoom(2)
	builder.BuildDoor(1, 2)

	return builder.GetMaze()
}

type WallFunc func() Wall
type RoomFunc func(n int) Room
type DoorFunc func(r1 Room, r2 Room) Door

type MazeGame struct {
	makeWall WallFunc
	makeRoom RoomFunc
	makeDoor DoorFunc
}

func (m MazeGame) Create() *Maze {
	mz := createMaze()

	r1 := m.makeRoom(1)
	r2 := m.makeRoom(2)
	mz.AddRoom(r1)
	mz.AddRoom(r2)

	d := m.makeDoor(r1, r2)

	r1.SetSide(DirectionNorth, m.makeWall())
	r1.SetSide(DirectionEast, d)
	r1.SetSide(DirectionSouth, m.makeWall())
	r1.SetSide(DirectionWest, m.makeWall())

	r2.SetSide(DirectionNorth, m.makeWall())
	r2.SetSide(DirectionEast, m.makeWall())
	r2.SetSide(DirectionSouth, m.makeWall())
	r2.SetSide(DirectionWest, d)

	return mz
}

func MakeSimpleGame() *MazeGame {
	return &MazeGame{makeDoor: CreateSimpleDoor, makeRoom: CreateSimpleRoom, makeWall: CreateSimpleWall}
}

func MakeBombedGame() *MazeGame {
	return &MazeGame{makeDoor: CreateSimpleDoor, makeRoom: createRoomWithBomb, makeWall: CreateBombedWall}
}

func MakeEnchantedGame() *MazeGame {
	return &MazeGame{makeDoor: CreateSpellDoor, makeRoom: CreateEnchantedRoom, makeWall: CreateSimpleWall}
}
